### PR TITLE
[Enhancement] Optimize global lock when creating table/mv/view

### PR DESCRIPTION
**Pull request by nshangyiming**
## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [x] 3.0
  - [x] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 59 / 86 (68.60%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|0|3|00.00%|[1039, 1040, 1041]|
|:large_blue_circle:|com/starrocks/catalog/Database.java|3|11|27.27%|[936, 937, 940, 941, 944, 945, 948, 949]|
|:large_blue_circle:|com/starrocks/common/util/concurrent/QueryableReentrantLock.java|3|5|60.00%|[30, 31]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|26|36|72.22%|[484, 2184, 2213, 2223, 2224, 2225, 2226, 2228, 2231, 2232]|
|:large_blue_circle:|com/starrocks/common/util/concurrent/MarkedCountDownLatch.java|17|20|85.00%|[69, 70, 72]|
|:large_blue_circle:|com/starrocks/common/util/StringUtils.java|7|8|87.50%|[39]|
|:large_blue_circle:|com/starrocks/common/util/concurrent/QueryableReentrantReadWriteLock.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 58 / 73 (79.45%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|25|36|69.44%|[484, 487, 2184, 2213, 2223, 2224, 2225, 2226, 2228, 2231, 2232]|
|:large_blue_circle:|com/starrocks/common/util/concurrent/MarkedCountDownLatch.java|17|20|85.00%|[69, 70, 72]|
|:large_blue_circle:|com/starrocks/common/util/StringUtils.java|7|8|87.50%|[39]|
|:large_blue_circle:|com/starrocks/catalog/Database.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/concurrent/QueryableReentrantReadWriteLock.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/concurrent/QueryableReentrantLock.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by nshangyiming** [comment]
@mergify rebase

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 62 / 73 (84.93%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|29|36|80.56%|[484, 487, 2184, 2213, 2228, 2231, 2232]|
|:large_blue_circle:|com/starrocks/common/util/concurrent/MarkedCountDownLatch.java|17|20|85.00%|[69, 70, 72]|
|:large_blue_circle:|com/starrocks/common/util/StringUtils.java|7|8|87.50%|[39]|
|:large_blue_circle:|com/starrocks/catalog/Database.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/concurrent/QueryableReentrantReadWriteLock.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/concurrent/QueryableReentrantLock.java|3|3|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/catalog/Database.java [hunk 1]
```diff
@@ -121,14 +121,21 @@ public class Database extends MetaObject implements Writable {
...
 
     // For external database location like hdfs://name_node:9000/user/hive/warehouse/test.db/
     private String location;
 
+    /**
+     * The number of operations that is modifying metadata of the database concurrently,
+     * like creating table or mv. When it's greater than 0, the database cannot be dropped.
+     * This is a runtime state and doesn't need to persist.
+     */
+    private CountingLatch runningMetaOps = new CountingLatch(0);
```
Comment (gengjun-git) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Database.java:137]: Where to update `runningMetaOps` ?

File: fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java [hunk 2]
```diff
@@ -463,39 +463,43 @@ public void replayCreateDb(CreateDbInfo createDbInfo) {
...
-                                    "The database [" + dbName +
-                                    "] cannot be dropped. If you want to forcibly drop(cannot be recovered)," +
-                                    " please use \"DROP DATABASE <database> FORCE\".");
-                }
+        List<Runnable> runnableList;
+        // 2. drop tables in db
+        db.writeLock();
+        try {
+            if (!db.isExist()) {
+                throw new MetaNotFoundException("Database not found");
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:484]: Add the database name into the exception to better trace.

File: fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java [hunk 3]
```diff
@@ -2173,51 +2177,71 @@ void setLakeStorageInfo(OlapTable table, String storageVolumeId, Map<String, Str
...
+        } finally {
+            unlock();
+        }
 
-            if (db.isSystemDatabase()) {
-                ErrorReport.reportDdlException(ErrorCode.ERR_CANT_CREATE_TABLE, table.getName(),
-                        "cannot create table in system database");
-            }
+        if (db.isSystemDatabase()) {
+            ErrorReport.reportDdlException(ErrorCode.ERR_CANT_CREATE_TABLE, table.getName(),
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:2213]: Add the database name in the exception

File: fe/fe-core/src/main/java/com/starrocks/alter/LakeTableAlterMetaJob.java [hunk 4]
```diff
@@ -31,10 +31,10 @@
 import com.starrocks.common.Config;
 import com.starrocks.common.DdlException;
 import com.starrocks.common.FeConstants;
-import com.starrocks.common.MarkedCountDownLatch;
 import com.starrocks.common.Pair;
 import com.starrocks.common.io.Text;
 import com.starrocks.common.util.PropertyAnalyzer;
+import com.starrocks.common.util.concurrent.MarkedCountDownLatch;
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/alter/LakeTableAlterMetaJob.java:37]: MarkedCountDownLatch seems not useful?

File: fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java (hunk 3 above)
Comment (nshangyiming) [reply fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:2213]: it's the system database, already reported...

File: fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java (hunk 2 above)
Comment (nshangyiming) [reply fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:484]: done

File: fe/fe-core/src/main/java/com/starrocks/alter/LakeTableAlterMetaJob.java (hunk 4 above)
Comment (nshangyiming) [reply fe/fe-core/src/main/java/com/starrocks/alter/LakeTableAlterMetaJob.java:37]: useful, check the code, besides i dont change the code, just put all the lock impl stuff in a same package

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 62 / 73 (84.93%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|29|36|80.56%|[484, 487, 2184, 2213, 2228, 2231, 2232]|
|:large_blue_circle:|com/starrocks/common/util/concurrent/MarkedCountDownLatch.java|17|20|85.00%|[69, 70, 72]|
|:large_blue_circle:|com/starrocks/common/util/StringUtils.java|7|8|87.50%|[39]|
|:large_blue_circle:|com/starrocks/catalog/Database.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/concurrent/QueryableReentrantReadWriteLock.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/concurrent/QueryableReentrantLock.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 60 / 71 (84.51%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|29|36|80.56%|[484, 487, 2184, 2213, 2228, 2231, 2232]|
|:large_blue_circle:|com/starrocks/common/util/concurrent/MarkedCountDownLatch.java|17|20|85.00%|[69, 70, 72]|
|:large_blue_circle:|com/starrocks/common/util/StringUtils.java|7|8|87.50%|[39]|
|:large_blue_circle:|com/starrocks/catalog/Database.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/concurrent/QueryableReentrantReadWriteLock.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/concurrent/QueryableReentrantLock.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by gengjun-git — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.2

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Pull request merged**